File: 2021/08/19/main.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func maxProduct(root *TreeNode) int {
	r := 0
	m := 1000000007
	s := []int{}
	var dfs func(*TreeNode) int

	dfs = func(root *TreeNode) int {
		if root == nil {
			return 0
		}
		s = append(s, root.Val+dfs(root.Left)+dfs(root.Right))
		return s[len(s)-1]
	}

	t := dfs(root)

	for _, x := range s {
		r = max(r, x*(t-x))
	}
	return r % m
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: 2021/08/19/main.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxProduct(self, root: Optional[TreeNode]) -> int:
        m = 10**9 + 7
        s = []
        def dfs(node, s):
            if not node: return 0
            s.append(node.val + dfs(node.left, s) + dfs(node.right, s))
            return s[-1]

        t = dfs(root, s)
        return max((x * (t - x) for x in s)) % m